design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,137 @@
+# ObjectBucketClaim CRD and Controller
+
+## Proposed Feature
+
+**A generalized method of S3 bucket provisioning through the implementation of an `ObjectBucketClaim` CRD and controller.** 
+
+Users will request buckets via the Kubernetes API and have returned ConfigMaps with connection information for that bucket.  
+
+One ConfigMap will be created per `ObjectBucketClaim` instance.  ConfigMap names will be deterministic and derived from the `ObjectBucketClaim` name.  This gives users the opportunity to define their Pod spec at the same time as the `ObjectBucketClaim`.  Additionally, the `ObjectBucketClaim` and Pod can be deployed at the same time thanks in-built synchronization in Kubernetes.
```

Comment:
Good point, will add pattern for naming

Reply:
I have a feeling we over designing this a bit. Pretty much all object storage systems have to handle dynamic bucket creation via AWS S3 API call. If they don't - then it means it is incompatible, and we should not bother. Rook framework would need to provide v2 and v4 authentication support, allow a user to select that and initiate the HTTP(S) call using CRD provided authorization credentials. Why do we need to introduce a storage class?

Reply:
That's true that objects storage systems should generally provide a standard API for bucket creation, but that's not the case for user creation (both radosgw and minio do it via a CLI, for example), and the provisioner needs to do both users and buckets.

The purpose of the StorageClass is to describe *which* object store to provision against.  It might be ceph cluster foo, AWS S3 region us-east, whatever.

Reply:
The storage class contains a provisioner property which can be used by the operator to decide how to create the bucket. Also the parameters property names the object store service name and namespace, which relieves the app developer from these concerns.

Reply:
>  a standard API for bucket creation, but that's not the case for user creation (both radosgw 
and minio do it via a CLI, for example), and the provisioner needs to do both users and buckets.

@liewegas Is the implication that bucket creation would be implemented redundantly in each rook object provisioner?  It's clear why we need to do this for users however bucket creation could be handled by a single operator with a AWS S3 client.

This is outside the scope of this proposal but it would be useful I think to implement a 2nd layer of abstraction for user provisioning with generalized user crd.  This general user CRD would also reference the same storageClass as proposed bucket CRD.  By parsing the storageClass, the user operator would create user CRDs for the respective object store operator.  E.g. `generalUser.spec.storagClass: rook-ceph.io` triggers the creation of `cephObjectStoreUser`, triggering the provisioning of the actual user.